From stable+bounces-200760-greg=kroah.com@vger.kernel.org Thu Dec 11 03:35:48 2025
From: Jingbo Xu <jefflexu@linux.alibaba.com>
Date: Thu, 11 Dec 2025 10:35:07 +0800
Subject: mm: fix arithmetic for max_prop_frac when setting max_ratio
To: gregkh@linuxfoundation.org, sashal@kernel.org, stable@vger.kernel.org
Cc: jefflexu@linux.alibaba.com
Message-ID: <20251211023507.82177-3-jefflexu@linux.alibaba.com>

From: Jingbo Xu <jefflexu@linux.alibaba.com>

commit fa151a39a6879144b587f35c0dfcc15e1be9450f upstream.

Since now bdi->max_ratio is part per million, fix the wrong arithmetic for
max_prop_frac when setting max_ratio.  Otherwise the miscalculated
max_prop_frac will affect the incrementing of writeout completion count
when max_ratio is not 100%.

Link: https://lkml.kernel.org/r/20231219142508.86265-3-jefflexu@linux.alibaba.com
Fixes: efc3e6ad53ea ("mm: split off __bdi_set_max_ratio() function")
Signed-off-by: Jingbo Xu <jefflexu@linux.alibaba.com>
Cc: Joseph Qi <joseph.qi@linux.alibaba.com>
Cc: Matthew Wilcox (Oracle) <willy@infradead.org>
Cc: Stefan Roesch <shr@devkernel.io>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
---
 mm/page-writeback.c |    3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

--- a/mm/page-writeback.c
+++ b/mm/page-writeback.c
@@ -750,7 +750,8 @@ static int __bdi_set_max_ratio(struct ba
 		ret = -EINVAL;
 	} else {
 		bdi->max_ratio = max_ratio;
-		bdi->max_prop_frac = (FPROP_FRAC_BASE * max_ratio) / 100;
+		bdi->max_prop_frac = (FPROP_FRAC_BASE * max_ratio) /
+						(100 * BDI_RATIO_SCALE);
 	}
 	spin_unlock_bh(&bdi_lock);
 
